pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/xpath/Attribute.java

```diff
@@ -48,8 +48,8 @@ public Object getValue() {
             return value;
         }
 
-        if (method.isAnnotationPresent(Deprecated.class) && LOG.isLoggable(Level.WARNING) && !DETECTED_DEPRECATED_ATTRIBUTES.contains(getLoggableAttributeName())) {
-            DETECTED_DEPRECATED_ATTRIBUTES.add(getLoggableAttributeName());
+        if (method.isAnnotationPresent(Deprecated.class) && LOG.isLoggable(Level.WARNING) && !DETECTED_DEPRECATED_ATTRIBUTES.containsKey(getLoggableAttributeName())) {
```

Comment:
if we change the type of `DETECTED_DEPRECATED_ATTRIBUTES` to keep it as `ConcurrentMap` we can change this non-atomic `containsKey` + `put` with a `putIfAbsent(getLoggableAttributeName(), Boolean.TRUE) == null`, which will be both thread-safe AND avoid race conditions that might get this logged twice for the same attribute